rfcs/2025-03-16-smp-queues.md

```diff
@@ -0,0 +1,163 @@
+
+Simply designating queue types would allow to use this information to decide for how long to retain queues, and potentially extending it:
+- unsecured 1-time invitation queues with sndSecure (support of securing by sender) - e.g., 3 months.
+- contact address queues without sndSecure - e.g., 3 years without activity.
+- Possibly, "queues" that prohibit messages and used only as blob storage - they would be used to store group profiles and superpeer addresses for the group.
+
+This proposal also combines NEW and NKEY command to streamline notifications in preparation to reworking of the notifications protocol.
+
+## Solution
+
+Current NEW and NKEY commands:
+
+```haskell
+NEW :: RcvPublicAuthKey -> RcvPublicDhKey -> Maybe BasicAuth -> SubscriptionMode -> SenderCanSecure -> Command Recipient
+
+NKEY :: NtfPublicAuthKey -> RcvNtfPublicDhKey -> Command Recipient
+
+-- | Queue IDs and keys, returned in IDS response
+data QueueIdsKeys = QIK
+  { rcvId :: RecipientId,
+    sndId :: SenderId,
+    rcvPublicDhKey :: RcvPublicDhKey,
+    sndSecure :: SenderCanSecure
+  }
+```
+
+Proposed NEW command replaces SenderCanSecure with QueueType, adds LinkId and SenderId, and combines NKEY command:
+
+```haskell
+NEW :: NewQueueRequest -> Command Recipient
+
+data NewQueueRequest = NewQueueRequest
+  { rcvAuthKey :: RcvPublicAuthKey,
+    rcvDhKey :: RcvPublicDhKey,
+    basicAuth :: Maybe BasicAuth,
+    subMode :: SubscriptionMode,
+    connMode :: ConnectionMode,
+    ntfRequest :: Maybe NtfRequest,
+    queueLink :: Maybe QueueLink
+  }
+
+-- An existing type.
+-- CMInvitation implies that sender can secure the queue.
+data ConnectionMode = CMInvitation | CMContact
+
+-- replaces NKEY command, includes NotifierId, it will be DRG-generated client-side.
+data NtfRequest = NtfRequest NotifierId NtfPublicAuthKey RcvNtfPublicDhKey
+
+-- LinkId is derived from private key passed to accepting party (sender) in the link, see below.
+-- SenderId is DRG-generated.
+data QueueLink = QueueLink LinkId SenderId EncLinkData
+
+newtype EncLinkData = EncLinkData ByteString -- encrypted link data
+
+data LinkData = LinkData
+  { connReq :: AConnectionRequestUri,
+    -- We need to use binary encoding for AConnectionRequestUri to reduce its size
+    -- connReq including the full link allows connection redundancy.
+    -- Technically, it would allow clients to point to queue on another server.
+    -- But clients should reject addresses when the primary queue is on a different server.
+    -- Clients could also warn when any of the queues is on unknown operator.
+    userData :: ByteString -- the max size needs to be estimated, but it is likely to be ~ 14kb
+  }
+
+-- | Updated queue IDs and keys, returned in IDS response
+data QueueIdsKeys = QIK
+  { rcvId :: RecipientId,
+    linkId :: Maybe LinkId, -- same as in ShortLink
+    sndId :: SenderId, -- same as in ShortLink or random
```

Comment:
why do these need to be returned?

Reply:
so that the difference is contained in encoding layer, and not that it sends completely different responses.